<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Roberts Lab Course Set - Things u should know - NCBI Blast</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Roberts Lab Course Set - Things u should know</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
    <a href="https://github.com/RobertsLab/tusk" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../modules/04-blast.html">NCBI Blast</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../modules/00-markdown.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">markdown</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../modules/01-github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">GitHub</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../modules/02-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">bash</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../modules/03-rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">RStudio</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../modules/04-blast.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">NCBI Blast</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tldr-basics" id="toc-tldr-basics" class="nav-link active" data-scroll-target="#tldr-basics">TLDR (basics)</a></li>
  <li><a href="#database-creation" id="toc-database-creation" class="nav-link" data-scroll-target="#database-creation">Database Creation</a>
  <ul class="collapse">
  <li><a href="#obtain-fasta-uniprotswiss-prot" id="toc-obtain-fasta-uniprotswiss-prot" class="nav-link" data-scroll-target="#obtain-fasta-uniprotswiss-prot">Obtain Fasta (UniProt/Swiss-Prot)</a></li>
  <li><a href="#making-the-database" id="toc-making-the-database" class="nav-link" data-scroll-target="#making-the-database">Making the database</a></li>
  </ul></li>
  <li><a href="#getting-the-query-fasta-file" id="toc-getting-the-query-fasta-file" class="nav-link" data-scroll-target="#getting-the-query-fasta-file">Getting the query fasta file</a></li>
  <li><a href="#running-blastx" id="toc-running-blastx" class="nav-link" data-scroll-target="#running-blastx">Running Blastx</a></li>
  <li><a href="#joining-blast-table-with-annoations." id="toc-joining-blast-table-with-annoations." class="nav-link" data-scroll-target="#joining-blast-table-with-annoations.">Joining Blast table with annoations.</a>
  <ul class="collapse">
  <li><a href="#prepping-blast-table-for-easy-join" id="toc-prepping-blast-table-for-easy-join" class="nav-link" data-scroll-target="#prepping-blast-table-for-easy-join">Prepping Blast table for easy join</a></li>
  <li><a href="#could-do-some-cool-stuff-in-r-here-reading-in-table" id="toc-could-do-some-cool-stuff-in-r-here-reading-in-table" class="nav-link" data-scroll-target="#could-do-some-cool-stuff-in-r-here-reading-in-table">Could do some cool stuff in R here reading in table</a></li>
  </ul></li>
  <li><a href="#case-study---oyster" id="toc-case-study---oyster" class="nav-link" data-scroll-target="#case-study---oyster">Case Study - Oyster</a>
  <ul class="collapse">
  <li><a href="#navigating-annotation" id="toc-navigating-annotation" class="nav-link" data-scroll-target="#navigating-annotation">Navigating Annotation</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">NCBI Blast</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
<p class="subtitle lead">Taking a set of unknown sequence files and annotating them</p>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>For the first task you will take an unknown multi-fasta file and annotate it using blast. You are welcome to do this in terminal, Rstudio, or jupyter. My recommendation, and how I will demonstrate is using Rmarkdown. Once you have have your project structured, we will download software, databases, a fasta file and run the code.</p>
<p>This is product offers a workflow to take a few thousand unidentified sequences and provide a better understanding of what genes are present. This will be accomplished through using Blast and protein sequenes from UniProt/Swiss-prot.</p>
<hr>
<section id="tldr-basics" class="level1">
<h1>TLDR (basics)</h1>
<pre><code>
/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb \
-in ../data/uniprot_sprot_r2023_01.fasta \
-dbtype prot \
-out ../blastdb/uniprot_sprot_r2023_01</code></pre>
<pre><code>/home/shared/ncbi-blast-2.11.0+/bin/blastx \
-query ../data/Ab_4denovo_CLC6_a.fa \
-db ../blastdb/uniprot_sprot_r2023_01 \
-out ../output/Ab_4-uniprot_blastx.tab \
-evalue 1E-20 \
-num_threads 20 \
-max_target_seqs 1 \
-outfmt 6</code></pre>
<p>A few weeks ago I perfected software installation, so I will not demonstrate that here. Please see this notebook for more.</p>
</section>
<section id="database-creation" class="level1">
<h1>Database Creation</h1>
<section id="obtain-fasta-uniprotswiss-prot" class="level2">
<h2 class="anchored" data-anchor-id="obtain-fasta-uniprotswiss-prot">Obtain Fasta (UniProt/Swiss-Prot)</h2>
<p>This is from here picur reviewe sequences I named based on the identify of the database given</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>current_time <span class="ot">&lt;-</span> <span class="fu">format</span>(<span class="fu">Sys.time</span>(), <span class="st">"%B %d, %Y %H:%M:%S"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"current date and time is "</span>, current_time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../data</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> uniprot_sprot.fasta.gz uniprot_sprot_r2023_04.fasta.gz</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> <span class="at">-k</span> uniprot_sprot_r2023_04.fasta.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="making-the-database" class="level2">
<h2 class="anchored" data-anchor-id="making-the-database">Making the database</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ../blastdb</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb</span> <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>-in ../data/uniprot_sprot_r2023_01.fasta <span class="dt">\</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>-dbtype prot <span class="dt">\</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>-out ../blastdb/uniprot_sprot_r2023_01</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="getting-the-query-fasta-file" class="level1">
<h1>Getting the query fasta file</h1>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> https://eagle.fish.washington.edu/cnidarian/Ab_4denovo_CLC6_a.fa <span class="dt">\</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>-k <span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> ../data/Ab_4denovo_CLC6_a.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
 20 1982k   20  403k    0     0   302k      0  0:00:06  0:00:01  0:00:05  303k
 57 1982k   57 1139k    0     0   478k      0  0:00:04  0:00:02  0:00:02  479k
 83 1982k   83 1659k    0     0   513k      0  0:00:03  0:00:03 --:--:--  514k
100 1982k  100 1982k    0     0   522k      0  0:00:03  0:00:03 --:--:--  523k</code></pre>
</div>
</div>
<p>Exploring what fasta file</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-3</span> ../data/Ab_4denovo_CLC6_a.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed_contig_1
ACACCCCACCCCAACGCACCCTCACCCCCACCCCAACAATCCATGATTGAATACTTCATC
TATCCAAGACAAACTCCTCCTACAATCCATGATAGAATTCCTCCAAAAATAATTTCACAC</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"How many sequences are there?"</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"&gt;"</span> ../data/Ab_4denovo_CLC6_a.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>How many sequences are there?
5490</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read FASTA file</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>fasta_file <span class="ot">&lt;-</span> <span class="st">"../data/Ab_4denovo_CLC6_a.fa"</span>  <span class="co"># Replace with the name of your FASTA file</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>sequences <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(fasta_file)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate sequence lengths</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>sequence_lengths <span class="ot">&lt;-</span> <span class="fu">width</span>(sequences)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a data frame</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>sequence_lengths_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Length =</span> sequence_lengths)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot histogram using ggplot2</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sequence_lengths_df, <span class="fu">aes</span>(<span class="at">x =</span> Length)) <span class="sc">+</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"grey"</span>, <span class="at">fill =</span> <span class="st">"blue"</span>, <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Histogram of Sequence Lengths"</span>,</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Sequence Length"</span>,</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Frequency"</span>) <span class="sc">+</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="04-blast_files/figure-html/histogram-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read FASTA file</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>fasta_file <span class="ot">&lt;-</span> <span class="st">"../data/Ab_4denovo_CLC6_a.fa"</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>sequences <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(fasta_file)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate base composition</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>base_composition <span class="ot">&lt;-</span> <span class="fu">alphabetFrequency</span>(sequences, <span class="at">baseOnly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame and reshape for ggplot2</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>base_composition_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(base_composition)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>base_composition_df<span class="sc">$</span>ID <span class="ot">&lt;-</span> <span class="fu">rownames</span>(base_composition_df)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>base_composition_melted <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(base_composition_df, <span class="at">id.vars =</span> <span class="st">"ID"</span>, <span class="at">variable.name =</span> <span class="st">"Base"</span>, <span class="at">value.name =</span> <span class="st">"Count"</span>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot base composition bar chart using ggplot2</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(base_composition_melted, <span class="fu">aes</span>(<span class="at">x =</span> Base, <span class="at">y =</span> Count, <span class="at">fill =</span> Base)) <span class="sc">+</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">position =</span> <span class="st">"dodge"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Base Composition"</span>,</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Base"</span>,</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"A"</span> <span class="ot">=</span> <span class="st">"green"</span>, <span class="st">"C"</span> <span class="ot">=</span> <span class="st">"blue"</span>, <span class="st">"G"</span> <span class="ot">=</span> <span class="st">"yellow"</span>, <span class="st">"T"</span> <span class="ot">=</span> <span class="st">"red"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="04-blast_files/figure-html/ACGT-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read FASTA file</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>fasta_file <span class="ot">&lt;-</span> <span class="st">"../data/Ab_4denovo_CLC6_a.fa"</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>sequences <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(fasta_file)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Count CG motifs in each sequence</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>count_cg_motifs <span class="ot">&lt;-</span> <span class="cf">function</span>(sequence) {</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  cg_motif <span class="ot">&lt;-</span> <span class="st">"CG"</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">length</span>(<span class="fu">gregexpr</span>(cg_motif, sequence, <span class="at">fixed =</span> <span class="cn">TRUE</span>)[[<span class="dv">1</span>]]))</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>cg_motifs_counts <span class="ot">&lt;-</span> <span class="fu">sapply</span>(sequences, count_cg_motifs)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a data frame</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>cg_motifs_counts_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">CG_Count =</span> cg_motifs_counts)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot CG motifs distribution using ggplot2</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(cg_motifs_counts_df, <span class="fu">aes</span>(<span class="at">x =</span> CG_Count)) <span class="sc">+</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="st">"blue"</span>, <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Distribution of CG Motifs"</span>,</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Number of CG Motifs"</span>,</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Frequency"</span>) <span class="sc">+</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="04-blast_files/figure-html/cg-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="running-blastx" class="level1">
<h1>Running Blastx</h1>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">~/applications/ncbi-blast-2.13.0+/bin/blastx</span> <span class="dt">\</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>-query ../data/Ab_4denovo_CLC6_a.fa <span class="dt">\</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>-db ../blastdb/uniprot_sprot_r2023_01 <span class="dt">\</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>-out ../output/Ab_4-uniprot_blastx.tab <span class="dt">\</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>-evalue 1E-20 <span class="dt">\</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>-num_threads 20 <span class="dt">\</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>-max_target_seqs 1 <span class="dt">\</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>-outfmt 6</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-2</span> ../output/Ab_4-uniprot_blastx.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Number of lines in output"</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wc</span> <span class="at">-l</span> ../output/Ab_4-uniprot_blastx.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="joining-blast-table-with-annoations." class="level1">
<h1>Joining Blast table with annoations.</h1>
<section id="prepping-blast-table-for-easy-join" class="level2">
<h2 class="anchored" data-anchor-id="prepping-blast-table-for-easy-join">Prepping Blast table for easy join</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tr</span> <span class="st">'|'</span> <span class="st">'\t'</span> <span class="op">&lt;</span> ../output/Ab_4-uniprot_blastx.tab <span class="dt">\</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> ../output/Ab_4-uniprot_blastx_sep.tab</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-1</span> ../output/Ab_4-uniprot_blastx_sep.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="could-do-some-cool-stuff-in-r-here-reading-in-table" class="level2">
<h2 class="anchored" data-anchor-id="could-do-some-cool-stuff-in-r-here-reading-in-table">Could do some cool stuff in R here reading in table</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>bltabl <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../output/Ab_4-uniprot_blastx_sep.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>spgo <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"https://gannet.fish.washington.edu/seashell/snaps/uniprot_table_r2023_01.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(<span class="fu">head</span>(bltabl), <span class="at">options =</span> <span class="fu">list</span>(<span class="at">scrollX =</span> <span class="cn">TRUE</span>, <span class="at">scrollY =</span> <span class="st">"400px"</span>, <span class="at">scrollCollapse =</span> <span class="cn">TRUE</span>, <span class="at">paging =</span> <span class="cn">FALSE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(<span class="fu">head</span>(spgo), <span class="at">options =</span> <span class="fu">list</span>(<span class="at">scrollX =</span> <span class="cn">TRUE</span>, <span class="at">scrollY =</span> <span class="st">"400px"</span>, <span class="at">scrollCollapse =</span> <span class="cn">TRUE</span>, <span class="at">paging =</span> <span class="cn">FALSE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(bltabl, spgo,  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"V3"</span> <span class="ot">=</span> <span class="st">"Entry"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">V1 =</span> <span class="fu">str_replace_all</span>(V1, </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">pattern =</span> <span class="st">"solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed"</span>, <span class="at">replacement =</span> <span class="st">"Ab"</span>))</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>annot_tab <span class="ot">&lt;-</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(bltabl, spgo,  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"V3"</span> <span class="ot">=</span> <span class="st">"Entry"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">V1 =</span> <span class="fu">str_replace_all</span>(V1, </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">pattern =</span> <span class="st">"solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed"</span>, <span class="at">replacement =</span> <span class="st">"Ab"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read dataset</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../output/blast_annot_go.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)  <span class="co"># Replace with the path to your dataset</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Select the column of interest</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>column_name <span class="ot">&lt;-</span> <span class="st">"Organism"</span>  <span class="co"># Replace with the name of the column of interest</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>column_data <span class="ot">&lt;-</span> dataset[[column_name]]</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Count the occurrences of the strings in the column</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>string_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(column_data)</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a data frame, sort by count, and select the top 10</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>string_counts_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(string_counts)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(string_counts_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"String"</span>, <span class="st">"Count"</span>)</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>string_counts_df <span class="ot">&lt;-</span> string_counts_df[<span class="fu">order</span>(string_counts_df<span class="sc">$</span>Count, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ]</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>top_10_strings <span class="ot">&lt;-</span> <span class="fu">head</span>(string_counts_df, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the top 10 most common strings using ggplot2</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(top_10_strings, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder</span>(String, <span class="sc">-</span>Count), <span class="at">y =</span> Count, <span class="at">fill =</span> String)) <span class="sc">+</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">position =</span> <span class="st">"dodge"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Top 10 Species hits"</span>,</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> column_name,</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../output/blast_annot_go.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename the `Gene.Ontology..biological.process.` column to `Biological_Process`</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(data)[<span class="fu">colnames</span>(data) <span class="sc">==</span> <span class="st">"Gene.Ontology..biological.process."</span>] <span class="ot">&lt;-</span> <span class="st">"Biological_Process"</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate the `Biological_Process` column into individual biological processes</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>data_separated <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(data<span class="sc">$</span>Biological_Process, <span class="at">split =</span> <span class="st">";"</span>))</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Trim whitespace from the biological processes</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>data_separated <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"^</span><span class="sc">\\</span><span class="st">s+|</span><span class="sc">\\</span><span class="st">s+$"</span>, <span class="st">""</span>, data_separated)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Count the occurrences of each biological process</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>process_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(data_separated)</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>process_counts <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Biological_Process =</span> <span class="fu">names</span>(process_counts), <span class="at">Count =</span> <span class="fu">as.integer</span>(process_counts))</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>process_counts <span class="ot">&lt;-</span> process_counts[<span class="fu">order</span>(<span class="sc">-</span>process_counts<span class="sc">$</span>Count), ]</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Select the 20 most predominant biological processes</span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>top_20_processes <span class="ot">&lt;-</span> process_counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, ]</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a color palette for the bars</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>bar_colors <span class="ot">&lt;-</span> <span class="fu">rainbow</span>(<span class="fu">nrow</span>(top_20_processes))</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a staggered vertical bar plot with different colors for each bar</span></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(top_20_processes<span class="sc">$</span>Count, <span class="at">names.arg =</span> <span class="fu">rep</span>(<span class="st">""</span>, <span class="fu">nrow</span>(top_20_processes)), <span class="at">col =</span> bar_colors,</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">max</span>(top_20_processes<span class="sc">$</span>Count) <span class="sc">*</span> <span class="fl">1.25</span>),</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Occurrences of the 20 Most Predominant Biological Processes"</span>, <span class="at">xlab =</span> <span class="st">"Biological Process"</span>, <span class="at">ylab =</span> <span class="st">"Count"</span>)</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a separate plot for the legend</span></span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="st">"../output/GOlegend.png"</span>, <span class="at">width =</span> <span class="dv">800</span>, <span class="at">height =</span> <span class="dv">600</span>)</span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))</span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.new</span>()</span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"center"</span>, <span class="at">legend =</span> top_20_processes<span class="sc">$</span>Biological_Process, <span class="at">fill =</span> bar_colors, <span class="at">cex =</span> <span class="dv">1</span>, <span class="at">title =</span> <span class="st">"Biological Processes"</span>)</span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"../output/GOlegend.png"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="case-study---oyster" class="level1">
<h1>Case Study - Oyster</h1>
<section id="navigating-annotation" class="level2">
<h2 class="anchored" data-anchor-id="navigating-annotation">Navigating Annotation</h2>
<p>The following is a stepwise example or annotation of a gene set using UniProt::Swiss-Prot (reviewed) such that Gene Ontology terms can be associated with each gene.</p>
<p>In this following chunk where the fasta file is downloaded the <a href="https://www.uniprot.org/help/release-statistics">release</a> is noted and the file name is modified accordingly.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> DRAFT_Funct_Enrich/annot</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> uniprot_sprot.fasta.gz uniprot_sprot_r2023_02.fasta.gz</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> <span class="at">-k</span> uniprot_sprot_r2023_02.fasta.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A protein blast database is then made.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb</span> <span class="dt">\</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>-in DRAFT_Funct_Enrich/annot/uniprot_sprot_r2023_02.fasta <span class="dt">\</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>-dbtype prot <span class="dt">\</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>-out DRAFT_Funct_Enrich/annot/uniprot_sprot_r2023_02</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In a majority of cases you will want to annotate a gene set to get gene ontology information. If you are creating your own genome or transcriptome it should be rather straightforward to know what file to annotate. If using a widely studied system where there are publically available resources, it is advisable to use those as this is the best way to facilitate integration of data sets. In this case study we will be considering the Eastern oyster, (<em>Crassostrea virginica</em>) for which there is data at <a href="https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_002022765.2/">NCBI</a> and <a href="http://metazoa.ensembl.org/Crassostrea_virginica_gca002022765v4/Info/Index">Ensembl Metazoa</a>. At NCBI there is both a GenBank and RefSeq assembly available.</p>
<p>In order to know which of the numerous fasta files should annotated with gene ontology information one should think downstream (or look to files already generated) to the identifiers in genesets that would be subject to functional enrichment tests.</p>
<p>The resulting fpkm count matrix for our case study is from an experiment where male and female oysters where exposed to low pH (and control) conditions. The count matrix is accessible <a href="https://github.com/epigeneticstoocean/2018_L18-adult-methylation/blob/main/analyses/gene_fpkm.csv">here (csv)</a>. Hisat2/Stringtie was used to generate the count matrix with <code>GCF_002022765.2_C_virginica-3.0_genomic.gff</code> formatting thus responsible for gene naming. Specifically the naming format is as follows <code>gene-LOC111099033,gene-LOC111099034,gene-LOC111099035</code>.</p>
<p>The following fasta was selected for annotation: <code>GCF_002022765.2_C_virginica-3.0_translated_cds.faa.gz</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> DRAFT_Funct_Enrich/annot</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/022/765/GCF_002022765.2_C_virginica-3.0/GCF_002022765.2_C_virginica-3.0_translated_cds.faa.gz</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> <span class="at">-k</span> GCF_002022765.2_C_virginica-3.0_translated_cds.faa.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-2</span> DRAFT_Funct_Enrich/annot/GCF_002022765.2_C_virginica-3.0_translated_cds.faa</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"number of sequences"</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span>  <span class="st">"&gt;"</span> DRAFT_Funct_Enrich/annot/GCF_002022765.2_C_virginica-3.0_translated_cds.faa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre><code>&gt;lcl|NC_035780.1_prot_XP_022327646.1_1 [gene=LOC111126949] [db_xref=GeneID:111126949] [protein=UNC5C-like protein] [protein_id=XP_022327646.1] [location=join(30535..31557,31736..31887,31977..32565,32959..33204)] [gbkey=CDS]
MTEVCYIWASSSTTVVICGIFFIVWRCFISIKKRASPLHGSSQQVCQTCQIEGHDFGEFQLSCRRQNTNVGYDLQGRRSD</code></pre>
<p>This protein fasta is used as query for blast of uniprot_sprot database.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/shared/ncbi-blast-2.11.0+/bin/blastp</span> <span class="dt">\</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>-query DRAFT_Funct_Enrich/annot/GCF_002022765.2_C_virginica-3.0_translated_cds.faa <span class="dt">\</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>-db DRAFT_Funct_Enrich/annot/uniprot_sprot_r2023_02 <span class="dt">\</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>-out DRAFT_Funct_Enrich/annot/Cvir_transcds-uniprot_blastp.tab <span class="dt">\</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>-evalue 1E-20 <span class="dt">\</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>-num_threads 40 <span class="dt">\</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>-max_target_seqs 1 <span class="dt">\</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>-outfmt 6</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is what the output file looks like, and at this point we want to get the UniProt Accession number for each gene</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-2</span> DRAFT_Funct_Enrich/annot/Cvir_transcds-uniprot_blastp.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>blast <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"DRAFT_Funct_Enrich/annot/Cvir_transcds-uniprot_blastp.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Convert fasta to tab</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">perl</span> <span class="at">-e</span> <span class="st">'$count=0; $len=0; while(&lt;&gt;) {s/\r?\n//; s/\t/ /g; if (s/^&gt;//) { if ($. != 1) {print "\n"} s/ |$/\t/; $count++; $_ .= "\t";} else {s/ //g; $len += length($_)} print $_;} print "\n"; warn "\nConverted $count FASTA records in $. lines to tabular format\nTotal sequence length: $len\n\n";'</span> <span class="dt">\</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>DRAFT_Funct_Enrich/annot/GCF_002022765.2_C_virginica-3.0_translated_cds.faa <span class="op">&gt;</span> DRAFT_Funct_Enrich/annot/GCF_002022765.2_C_virginica-3.0_translated_cds.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-1</span> DRAFT_Funct_Enrich/annot/GCF_002022765.2_C_virginica-3.0_translated_cds.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>cdsftab <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"DRAFT_Funct_Enrich/annot/GCF_002022765.2_C_virginica-3.0_translated_cds.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">row.names=</span><span class="cn">NULL</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can take the two data frames: A) blast output of taking protein fasta and comparing to uniprot_swiss-prot and B) a tabular version of same fasta file that has ID numbers of importance. Note this importance was determined based on what we want to use down stream.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>g.spid <span class="ot">&lt;-</span> <span class="fu">left_join</span>(blast, cdsftab, <span class="at">by =</span> <span class="st">"V1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">str_extract</span>(V2.y, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">[gene=)</span><span class="sc">\\</span><span class="st">w+"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene, V11, V2.x) <span class="sc">%&gt;%</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">SPID =</span> <span class="fu">str_extract</span>(V2.x, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">|)[^</span><span class="sc">\\</span><span class="st">|]*(?=</span><span class="sc">\\</span><span class="st">|)"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(gene, SPID, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lets break it down step by step:</p>
<ol type="1">
<li><p><strong><code>g.spid &lt;- left_join(blast, cdsftab, by = "V1")</code></strong> - This line is using the <strong><code>left_join()</code></strong> function from <strong><code>dplyr</code></strong> to merge the <strong><code>blast</code></strong> and <strong><code>cdsftab</code></strong> datasets by the column V1. A left join retains all the rows in the <strong><code>blast</code></strong> data frame and appends the matching rows in the <strong><code>cdsftab</code></strong> data frame. If there is no match, the result is <strong><code>NA</code></strong>. The result of this operation is assigned to the <strong><code>g.spid</code></strong> object.</p></li>
<li><p><strong><code>mutate(gene = str_extract(V2.y, "(?&lt;=\\[gene=)\\w+"))</code></strong> - This line is using the <strong><code>mutate()</code></strong> function from <strong><code>dplyr</code></strong> to add a new column called gene to the data frame. The new column is created by extracting substrings from the V2.y column based on the given regular expression pattern <strong><code>"(?&lt;=\\[gene=)\\w+"</code></strong>. This regular expression matches and extracts any word (sequence of word characters, i.e., alphanumeric and underscore) that comes after [gene=.</p></li>
<li><p><strong><code>select(gene, V11, V2.x)</code></strong> - This line is using the <strong><code>select()</code></strong> function from <strong><code>dplyr</code></strong> to keep only the specified columns (gene, V11, and V2.x) in the data frame.</p></li>
<li><p><strong><code>mutate(SPID = str_extract(V2.x, "(?&lt;=\\|)[^\\|]*(?=\\|)"))</code></strong> - Again, the <strong><code>mutate()</code></strong> function is used to add another new column named SPID. This column is created by extracting substrings from the V2.x column. The regular expression <strong><code>"(?&lt;=\\|)[^\\|]*(?=\\|)"</code></strong> is designed to extract any character(s) that is/are surrounded by | (pipe symbol). This is a common format for delimited strings.</p></li>
<li><p><strong><code>distinct(gene, SPID, .keep_all = TRUE)</code></strong> - This line is using the <strong><code>distinct()</code></strong> function from <strong><code>dplyr</code></strong> to remove duplicate rows based on the gene and SPID columns. The <strong><code>.keep_all = TRUE</code></strong> argument means that all other columns are also kept in the result, not just the gene and SPID columns.</p></li>
</ol>
<p>The resulting <strong><code>g.spid</code></strong> data frame should have unique rows with respect to the gene and SPID columns, and it should contain these two new columns, gene and SPID, extracted from the original data based on specific string patterns.</p>
<p>Now lets just write out SPIDs.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(blast, cdsftab, <span class="at">by =</span> <span class="st">"V1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">str_extract</span>(V2.y, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">[gene=)</span><span class="sc">\\</span><span class="st">w+"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene, V11, V2.x) <span class="sc">%&gt;%</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">SPID =</span> <span class="fu">str_extract</span>(V2.x, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">|)[^</span><span class="sc">\\</span><span class="st">|]*(?=</span><span class="sc">\\</span><span class="st">|)"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(gene, SPID, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(SPID) <span class="sc">%&gt;%</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(<span class="at">file =</span> <span class="st">"DRAFT_Funct_Enrich/annot/SPID.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a> ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With a list of matching Swiss-Prot IDs, (technically UniProt Accession number) we can go back to https://www.uniprot.org and grab corresponding GO terms. This can be done via a web or using Python API.</p>
<p><strong>Using Web</strong></p>
<p>Using ID Mapping</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="http://gannet.fish.washington.edu/seashell/snaps/Monosnap_RetrieveID_mapping__UniProt_2023-07-10_10-03-33.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">id</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="http://gannet.fish.washington.edu/seashell/snaps/Monosnap_Tool_results__UniProt_2023-07-10_10-05-07.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">finished</figcaption>
</figure>
</div>
<p>Now will customize columns to get GO IDs.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="http://gannet.fish.washington.edu/seashell/snaps/Monosnap_RetrieveID_mapping_results__UniProtKB__UniProt_2023-07-10_10-07-05.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">custcol</figcaption>
</figure>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> <span class="at">-2</span> DRAFT_Funct_Enrich/annot/uniprotGO.tab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally we can join table to get LOCIDs the notation for our DEGs, with GO terms.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>go <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"DRAFT_Funct_Enrich/annot/uniprotGO.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">row.names=</span><span class="cn">NULL</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(g.spid, go, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"SPID"</span> <span class="ot">=</span> <span class="st">"Entry"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene,Gene.Ontology.IDs) <span class="sc">%&gt;%</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(<span class="at">file =</span> <span class="st">"DRAFT_Funct_Enrich/annot/geneGO.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> DRAFT_Funct_Enrich/annot/geneGO.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Using API</strong></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python3</span> DRAFT_Funct_Enrich/annot/uniprot-retrieval.py DRAFT_Funct_Enrich/annot/SPID.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb45" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "NCBI Blast"</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="an">subtitle:</span><span class="co"> "Taking a set of unknown sequence files and annotating them"</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: false</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a><span class="co">    code-copy: true</span></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="co">    highlight-style: github</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a><span class="co">    code-overflow: wrap</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>For the first task you will take an unknown multi-fasta file and annotate it using blast. You are welcome to do this in terminal, Rstudio, or jupyter. My recommendation, and how I will demonstrate is using Rmarkdown. Once you have have your project structured, we will download software, databases, a fasta file and run the code.</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a><span class="in">```{r setup, include = FALSE}</span></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(knitr)</span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kableExtra)</span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DT)</span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(</span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">echo =</span> <span class="cn">TRUE</span>,         <span class="co"># Display code chunks</span></span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">eval =</span> <span class="cn">FALSE</span>,         <span class="co"># Evaluate code chunks</span></span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">warning =</span> <span class="cn">FALSE</span>,     <span class="co"># Hide warnings</span></span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">message =</span> <span class="cn">FALSE</span>,     <span class="co"># Hide messages</span></span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">fig.width =</span> <span class="dv">6</span>,       <span class="co"># Set plot width in inches</span></span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">fig.height =</span> <span class="dv">4</span>,      <span class="co"># Set plot height in inches</span></span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">fig.align =</span> <span class="st">"center"</span> <span class="co"># Align plots to the center</span></span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-31"><a href="#cb45-31" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-32"><a href="#cb45-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-33"><a href="#cb45-33" aria-hidden="true" tabindex="-1"></a>This is product offers a workflow to take a few thousand unidentified sequences and provide a better understanding of what genes are present. This will be accomplished through using Blast and protein sequenes from UniProt/Swiss-prot.</span>
<span id="cb45-34"><a href="#cb45-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-35"><a href="#cb45-35" aria-hidden="true" tabindex="-1"></a>------------------------------------------------------------------------</span>
<span id="cb45-36"><a href="#cb45-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-37"><a href="#cb45-37" aria-hidden="true" tabindex="-1"></a><span class="fu"># TLDR (basics)</span></span>
<span id="cb45-38"><a href="#cb45-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-39"><a href="#cb45-39" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-40"><a href="#cb45-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-41"><a href="#cb45-41" aria-hidden="true" tabindex="-1"></a><span class="in">/home/shared/ncbi-blast-2.11.0+/bin/makeblastdb \</span></span>
<span id="cb45-42"><a href="#cb45-42" aria-hidden="true" tabindex="-1"></a><span class="in">-in ../data/uniprot_sprot_r2023_01.fasta \</span></span>
<span id="cb45-43"><a href="#cb45-43" aria-hidden="true" tabindex="-1"></a><span class="in">-dbtype prot \</span></span>
<span id="cb45-44"><a href="#cb45-44" aria-hidden="true" tabindex="-1"></a><span class="in">-out ../blastdb/uniprot_sprot_r2023_01</span></span>
<span id="cb45-45"><a href="#cb45-45" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-46"><a href="#cb45-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-47"><a href="#cb45-47" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-48"><a href="#cb45-48" aria-hidden="true" tabindex="-1"></a><span class="in">/home/shared/ncbi-blast-2.11.0+/bin/blastx \</span></span>
<span id="cb45-49"><a href="#cb45-49" aria-hidden="true" tabindex="-1"></a><span class="in">-query ../data/Ab_4denovo_CLC6_a.fa \</span></span>
<span id="cb45-50"><a href="#cb45-50" aria-hidden="true" tabindex="-1"></a><span class="in">-db ../blastdb/uniprot_sprot_r2023_01 \</span></span>
<span id="cb45-51"><a href="#cb45-51" aria-hidden="true" tabindex="-1"></a><span class="in">-out ../output/Ab_4-uniprot_blastx.tab \</span></span>
<span id="cb45-52"><a href="#cb45-52" aria-hidden="true" tabindex="-1"></a><span class="in">-evalue 1E-20 \</span></span>
<span id="cb45-53"><a href="#cb45-53" aria-hidden="true" tabindex="-1"></a><span class="in">-num_threads 20 \</span></span>
<span id="cb45-54"><a href="#cb45-54" aria-hidden="true" tabindex="-1"></a><span class="in">-max_target_seqs 1 \</span></span>
<span id="cb45-55"><a href="#cb45-55" aria-hidden="true" tabindex="-1"></a><span class="in">-outfmt 6</span></span>
<span id="cb45-56"><a href="#cb45-56" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-57"><a href="#cb45-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-58"><a href="#cb45-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-59"><a href="#cb45-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-60"><a href="#cb45-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-61"><a href="#cb45-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-62"><a href="#cb45-62" aria-hidden="true" tabindex="-1"></a>A few weeks ago I perfected software installation, so I will not demonstrate that here. Please see this notebook for more.</span>
<span id="cb45-63"><a href="#cb45-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-64"><a href="#cb45-64" aria-hidden="true" tabindex="-1"></a><span class="fu"># Database Creation</span></span>
<span id="cb45-65"><a href="#cb45-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-66"><a href="#cb45-66" aria-hidden="true" tabindex="-1"></a><span class="fu">## Obtain Fasta (UniProt/Swiss-Prot)</span></span>
<span id="cb45-67"><a href="#cb45-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-68"><a href="#cb45-68" aria-hidden="true" tabindex="-1"></a>This is from here picur reviewe sequences I named based on the identify of the database given</span>
<span id="cb45-69"><a href="#cb45-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-70"><a href="#cb45-70" aria-hidden="true" tabindex="-1"></a><span class="in">```{r time, eval=FALSE, echo=TRUE}</span></span>
<span id="cb45-71"><a href="#cb45-71" aria-hidden="true" tabindex="-1"></a>current_time <span class="ot">&lt;-</span> <span class="fu">format</span>(<span class="fu">Sys.time</span>(), <span class="st">"%B %d, %Y %H:%M:%S"</span>)</span>
<span id="cb45-72"><a href="#cb45-72" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"current date and time is "</span>, current_time)</span>
<span id="cb45-73"><a href="#cb45-73" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-74"><a href="#cb45-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-75"><a href="#cb45-75" aria-hidden="true" tabindex="-1"></a><span class="in">```{r download-data, engine='bash'}</span></span>
<span id="cb45-76"><a href="#cb45-76" aria-hidden="true" tabindex="-1"></a>cd ..<span class="sc">/</span>data</span>
<span id="cb45-77"><a href="#cb45-77" aria-hidden="true" tabindex="-1"></a>curl <span class="sc">-</span>O https<span class="sc">:</span><span class="er">//</span>ftp.uniprot.org<span class="sc">/</span>pub<span class="sc">/</span>databases<span class="sc">/</span>uniprot<span class="sc">/</span>current_release<span class="sc">/</span>knowledgebase<span class="sc">/</span>complete<span class="sc">/</span>uniprot_sprot.fasta.gz</span>
<span id="cb45-78"><a href="#cb45-78" aria-hidden="true" tabindex="-1"></a>mv uniprot_sprot.fasta.gz uniprot_sprot_r2023_04.fasta.gz</span>
<span id="cb45-79"><a href="#cb45-79" aria-hidden="true" tabindex="-1"></a>gunzip <span class="sc">-</span>k uniprot_sprot_r2023_04.fasta.gz</span>
<span id="cb45-80"><a href="#cb45-80" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-81"><a href="#cb45-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-82"><a href="#cb45-82" aria-hidden="true" tabindex="-1"></a><span class="fu">## Making the database</span></span>
<span id="cb45-83"><a href="#cb45-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-84"><a href="#cb45-84" aria-hidden="true" tabindex="-1"></a><span class="in">```{r make-blastdb, engine='bash'}</span></span>
<span id="cb45-85"><a href="#cb45-85" aria-hidden="true" tabindex="-1"></a>mkdir ..<span class="sc">/</span>blastdb</span>
<span id="cb45-86"><a href="#cb45-86" aria-hidden="true" tabindex="-1"></a><span class="sc">/</span>home<span class="sc">/</span>shared<span class="sc">/</span>ncbi<span class="sc">-</span>blast<span class="dv">-2</span>.<span class="fl">11.0</span><span class="sc">+</span><span class="er">/</span>bin<span class="sc">/</span>makeblastdb \</span>
<span id="cb45-87"><a href="#cb45-87" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span><span class="cf">in</span> ..<span class="sc">/</span>data<span class="sc">/</span>uniprot_sprot_r2023_01.fasta \</span>
<span id="cb45-88"><a href="#cb45-88" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>dbtype prot \</span>
<span id="cb45-89"><a href="#cb45-89" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>out ..<span class="sc">/</span>blastdb<span class="sc">/</span>uniprot_sprot_r2023_01</span>
<span id="cb45-90"><a href="#cb45-90" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-91"><a href="#cb45-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-92"><a href="#cb45-92" aria-hidden="true" tabindex="-1"></a><span class="fu"># Getting the query fasta file</span></span>
<span id="cb45-93"><a href="#cb45-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-94"><a href="#cb45-94" aria-hidden="true" tabindex="-1"></a><span class="in">```{r download-query, engine='bash', eval=TRUE}</span></span>
<span id="cb45-95"><a href="#cb45-95" aria-hidden="true" tabindex="-1"></a>curl https<span class="sc">:</span><span class="er">//</span>eagle.fish.washington.edu<span class="sc">/</span>cnidarian<span class="sc">/</span>Ab_4denovo_CLC6_a.fa \</span>
<span id="cb45-96"><a href="#cb45-96" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>k \</span>
<span id="cb45-97"><a href="#cb45-97" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ..<span class="sc">/</span>data<span class="sc">/</span>Ab_4denovo_CLC6_a.fa</span>
<span id="cb45-98"><a href="#cb45-98" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-99"><a href="#cb45-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-100"><a href="#cb45-100" aria-hidden="true" tabindex="-1"></a>Exploring what fasta file</span>
<span id="cb45-101"><a href="#cb45-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-102"><a href="#cb45-102" aria-hidden="true" tabindex="-1"></a><span class="in">```{r view-query, eval = TRUE, engine='bash'}</span></span>
<span id="cb45-103"><a href="#cb45-103" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">3</span> ..<span class="sc">/</span>data<span class="sc">/</span>Ab_4denovo_CLC6_a.fa</span>
<span id="cb45-104"><a href="#cb45-104" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-105"><a href="#cb45-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-106"><a href="#cb45-106" aria-hidden="true" tabindex="-1"></a><span class="in">```{r view2-query, engine='bash', eval=TRUE}</span></span>
<span id="cb45-107"><a href="#cb45-107" aria-hidden="true" tabindex="-1"></a>echo <span class="st">"How many sequences are there?"</span></span>
<span id="cb45-108"><a href="#cb45-108" aria-hidden="true" tabindex="-1"></a>grep <span class="sc">-</span>c <span class="st">"&gt;"</span> ..<span class="sc">/</span>data<span class="sc">/</span>Ab_4denovo_CLC6_a.fa</span>
<span id="cb45-109"><a href="#cb45-109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-110"><a href="#cb45-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-111"><a href="#cb45-111" aria-hidden="true" tabindex="-1"></a><span class="in">```{r histogram, eval=TRUE}</span></span>
<span id="cb45-112"><a href="#cb45-112" aria-hidden="true" tabindex="-1"></a><span class="co"># Read FASTA file</span></span>
<span id="cb45-113"><a href="#cb45-113" aria-hidden="true" tabindex="-1"></a>fasta_file <span class="ot">&lt;-</span> <span class="st">"../data/Ab_4denovo_CLC6_a.fa"</span>  <span class="co"># Replace with the name of your FASTA file</span></span>
<span id="cb45-114"><a href="#cb45-114" aria-hidden="true" tabindex="-1"></a>sequences <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(fasta_file)</span>
<span id="cb45-115"><a href="#cb45-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-116"><a href="#cb45-116" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate sequence lengths</span></span>
<span id="cb45-117"><a href="#cb45-117" aria-hidden="true" tabindex="-1"></a>sequence_lengths <span class="ot">&lt;-</span> <span class="fu">width</span>(sequences)</span>
<span id="cb45-118"><a href="#cb45-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-119"><a href="#cb45-119" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a data frame</span></span>
<span id="cb45-120"><a href="#cb45-120" aria-hidden="true" tabindex="-1"></a>sequence_lengths_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Length =</span> sequence_lengths)</span>
<span id="cb45-121"><a href="#cb45-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-122"><a href="#cb45-122" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot histogram using ggplot2</span></span>
<span id="cb45-123"><a href="#cb45-123" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sequence_lengths_df, <span class="fu">aes</span>(<span class="at">x =</span> Length)) <span class="sc">+</span></span>
<span id="cb45-124"><a href="#cb45-124" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"grey"</span>, <span class="at">fill =</span> <span class="st">"blue"</span>, <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb45-125"><a href="#cb45-125" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Histogram of Sequence Lengths"</span>,</span>
<span id="cb45-126"><a href="#cb45-126" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Sequence Length"</span>,</span>
<span id="cb45-127"><a href="#cb45-127" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Frequency"</span>) <span class="sc">+</span></span>
<span id="cb45-128"><a href="#cb45-128" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb45-129"><a href="#cb45-129" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-130"><a href="#cb45-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-131"><a href="#cb45-131" aria-hidden="true" tabindex="-1"></a><span class="in">```{r ACGT, eval=TRUE}</span></span>
<span id="cb45-132"><a href="#cb45-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-133"><a href="#cb45-133" aria-hidden="true" tabindex="-1"></a><span class="co"># Read FASTA file</span></span>
<span id="cb45-134"><a href="#cb45-134" aria-hidden="true" tabindex="-1"></a>fasta_file <span class="ot">&lt;-</span> <span class="st">"../data/Ab_4denovo_CLC6_a.fa"</span></span>
<span id="cb45-135"><a href="#cb45-135" aria-hidden="true" tabindex="-1"></a>sequences <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(fasta_file)</span>
<span id="cb45-136"><a href="#cb45-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-137"><a href="#cb45-137" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate base composition</span></span>
<span id="cb45-138"><a href="#cb45-138" aria-hidden="true" tabindex="-1"></a>base_composition <span class="ot">&lt;-</span> <span class="fu">alphabetFrequency</span>(sequences, <span class="at">baseOnly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb45-139"><a href="#cb45-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-140"><a href="#cb45-140" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to data frame and reshape for ggplot2</span></span>
<span id="cb45-141"><a href="#cb45-141" aria-hidden="true" tabindex="-1"></a>base_composition_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(base_composition)</span>
<span id="cb45-142"><a href="#cb45-142" aria-hidden="true" tabindex="-1"></a>base_composition_df<span class="sc">$</span>ID <span class="ot">&lt;-</span> <span class="fu">rownames</span>(base_composition_df)</span>
<span id="cb45-143"><a href="#cb45-143" aria-hidden="true" tabindex="-1"></a>base_composition_melted <span class="ot">&lt;-</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(base_composition_df, <span class="at">id.vars =</span> <span class="st">"ID"</span>, <span class="at">variable.name =</span> <span class="st">"Base"</span>, <span class="at">value.name =</span> <span class="st">"Count"</span>)</span>
<span id="cb45-144"><a href="#cb45-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-145"><a href="#cb45-145" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot base composition bar chart using ggplot2</span></span>
<span id="cb45-146"><a href="#cb45-146" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(base_composition_melted, <span class="fu">aes</span>(<span class="at">x =</span> Base, <span class="at">y =</span> Count, <span class="at">fill =</span> Base)) <span class="sc">+</span></span>
<span id="cb45-147"><a href="#cb45-147" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">position =</span> <span class="st">"dodge"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb45-148"><a href="#cb45-148" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Base Composition"</span>,</span>
<span id="cb45-149"><a href="#cb45-149" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Base"</span>,</span>
<span id="cb45-150"><a href="#cb45-150" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span></span>
<span id="cb45-151"><a href="#cb45-151" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb45-152"><a href="#cb45-152" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"A"</span> <span class="ot">=</span> <span class="st">"green"</span>, <span class="st">"C"</span> <span class="ot">=</span> <span class="st">"blue"</span>, <span class="st">"G"</span> <span class="ot">=</span> <span class="st">"yellow"</span>, <span class="st">"T"</span> <span class="ot">=</span> <span class="st">"red"</span>))</span>
<span id="cb45-153"><a href="#cb45-153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-154"><a href="#cb45-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-155"><a href="#cb45-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-156"><a href="#cb45-156" aria-hidden="true" tabindex="-1"></a><span class="in">```{r cg, eval=TRUE}</span></span>
<span id="cb45-157"><a href="#cb45-157" aria-hidden="true" tabindex="-1"></a><span class="co"># Read FASTA file</span></span>
<span id="cb45-158"><a href="#cb45-158" aria-hidden="true" tabindex="-1"></a>fasta_file <span class="ot">&lt;-</span> <span class="st">"../data/Ab_4denovo_CLC6_a.fa"</span></span>
<span id="cb45-159"><a href="#cb45-159" aria-hidden="true" tabindex="-1"></a>sequences <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(fasta_file)</span>
<span id="cb45-160"><a href="#cb45-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-161"><a href="#cb45-161" aria-hidden="true" tabindex="-1"></a><span class="co"># Count CG motifs in each sequence</span></span>
<span id="cb45-162"><a href="#cb45-162" aria-hidden="true" tabindex="-1"></a>count_cg_motifs <span class="ot">&lt;-</span> <span class="cf">function</span>(sequence) {</span>
<span id="cb45-163"><a href="#cb45-163" aria-hidden="true" tabindex="-1"></a>  cg_motif <span class="ot">&lt;-</span> <span class="st">"CG"</span></span>
<span id="cb45-164"><a href="#cb45-164" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">length</span>(<span class="fu">gregexpr</span>(cg_motif, sequence, <span class="at">fixed =</span> <span class="cn">TRUE</span>)[[<span class="dv">1</span>]]))</span>
<span id="cb45-165"><a href="#cb45-165" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb45-166"><a href="#cb45-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-167"><a href="#cb45-167" aria-hidden="true" tabindex="-1"></a>cg_motifs_counts <span class="ot">&lt;-</span> <span class="fu">sapply</span>(sequences, count_cg_motifs)</span>
<span id="cb45-168"><a href="#cb45-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-169"><a href="#cb45-169" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a data frame</span></span>
<span id="cb45-170"><a href="#cb45-170" aria-hidden="true" tabindex="-1"></a>cg_motifs_counts_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">CG_Count =</span> cg_motifs_counts)</span>
<span id="cb45-171"><a href="#cb45-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-172"><a href="#cb45-172" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot CG motifs distribution using ggplot2</span></span>
<span id="cb45-173"><a href="#cb45-173" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(cg_motifs_counts_df, <span class="fu">aes</span>(<span class="at">x =</span> CG_Count)) <span class="sc">+</span></span>
<span id="cb45-174"><a href="#cb45-174" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="st">"blue"</span>, <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb45-175"><a href="#cb45-175" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Distribution of CG Motifs"</span>,</span>
<span id="cb45-176"><a href="#cb45-176" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Number of CG Motifs"</span>,</span>
<span id="cb45-177"><a href="#cb45-177" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Frequency"</span>) <span class="sc">+</span></span>
<span id="cb45-178"><a href="#cb45-178" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb45-179"><a href="#cb45-179" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-180"><a href="#cb45-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-181"><a href="#cb45-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-182"><a href="#cb45-182" aria-hidden="true" tabindex="-1"></a><span class="fu"># Running Blastx</span></span>
<span id="cb45-183"><a href="#cb45-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-184"><a href="#cb45-184" aria-hidden="true" tabindex="-1"></a><span class="in">```{r blastx, engine='bash'}</span></span>
<span id="cb45-185"><a href="#cb45-185" aria-hidden="true" tabindex="-1"></a><span class="sc">~</span><span class="er">/</span>applications<span class="sc">/</span>ncbi<span class="sc">-</span>blast<span class="dv">-2</span>.<span class="fl">13.0</span><span class="sc">+</span><span class="er">/</span>bin<span class="sc">/</span>blastx \</span>
<span id="cb45-186"><a href="#cb45-186" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>query ..<span class="sc">/</span>data<span class="sc">/</span>Ab_4denovo_CLC6_a.fa \</span>
<span id="cb45-187"><a href="#cb45-187" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>db ..<span class="sc">/</span>blastdb<span class="sc">/</span>uniprot_sprot_r2023_01 \</span>
<span id="cb45-188"><a href="#cb45-188" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>out ..<span class="sc">/</span>output<span class="sc">/</span>Ab_4<span class="sc">-</span>uniprot_blastx.tab \</span>
<span id="cb45-189"><a href="#cb45-189" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>evalue <span class="fl">1E-20</span> \</span>
<span id="cb45-190"><a href="#cb45-190" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>num_threads <span class="dv">20</span> \</span>
<span id="cb45-191"><a href="#cb45-191" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>max_target_seqs <span class="dv">1</span> \</span>
<span id="cb45-192"><a href="#cb45-192" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>outfmt <span class="dv">6</span></span>
<span id="cb45-193"><a href="#cb45-193" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-194"><a href="#cb45-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-195"><a href="#cb45-195" aria-hidden="true" tabindex="-1"></a><span class="in">```{r blast-look, engine='bash'}</span></span>
<span id="cb45-196"><a href="#cb45-196" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">2</span> ..<span class="sc">/</span>output<span class="sc">/</span>Ab_4<span class="sc">-</span>uniprot_blastx.tab</span>
<span id="cb45-197"><a href="#cb45-197" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-198"><a href="#cb45-198" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-199"><a href="#cb45-199" aria-hidden="true" tabindex="-1"></a><span class="in">```{r blast-look2, engine='bash'}</span></span>
<span id="cb45-200"><a href="#cb45-200" aria-hidden="true" tabindex="-1"></a>echo <span class="st">"Number of lines in output"</span></span>
<span id="cb45-201"><a href="#cb45-201" aria-hidden="true" tabindex="-1"></a>wc <span class="sc">-</span>l ..<span class="sc">/</span>output<span class="sc">/</span>Ab_4<span class="sc">-</span>uniprot_blastx.tab</span>
<span id="cb45-202"><a href="#cb45-202" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-203"><a href="#cb45-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-204"><a href="#cb45-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-205"><a href="#cb45-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-206"><a href="#cb45-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-207"><a href="#cb45-207" aria-hidden="true" tabindex="-1"></a><span class="fu"># Joining Blast table with annoations.</span></span>
<span id="cb45-208"><a href="#cb45-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-209"><a href="#cb45-209" aria-hidden="true" tabindex="-1"></a><span class="fu">## Prepping Blast table for easy join</span></span>
<span id="cb45-210"><a href="#cb45-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-211"><a href="#cb45-211" aria-hidden="true" tabindex="-1"></a><span class="in">```{r separate, engine='bash'}</span></span>
<span id="cb45-212"><a href="#cb45-212" aria-hidden="true" tabindex="-1"></a>tr <span class="st">'|'</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span> <span class="sc">&lt;</span> ..<span class="sc">/</span>output<span class="sc">/</span>Ab_4<span class="sc">-</span>uniprot_blastx.tab \</span>
<span id="cb45-213"><a href="#cb45-213" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ..<span class="sc">/</span>output<span class="sc">/</span>Ab_4<span class="sc">-</span>uniprot_blastx_sep.tab</span>
<span id="cb45-214"><a href="#cb45-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-215"><a href="#cb45-215" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">1</span> ..<span class="sc">/</span>output<span class="sc">/</span>Ab_4<span class="sc">-</span>uniprot_blastx_sep.tab</span>
<span id="cb45-216"><a href="#cb45-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-217"><a href="#cb45-217" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-218"><a href="#cb45-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-219"><a href="#cb45-219" aria-hidden="true" tabindex="-1"></a><span class="fu">## Could do some cool stuff in R here reading in table</span></span>
<span id="cb45-220"><a href="#cb45-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-221"><a href="#cb45-221" aria-hidden="true" tabindex="-1"></a><span class="in">```{r read-data}</span></span>
<span id="cb45-222"><a href="#cb45-222" aria-hidden="true" tabindex="-1"></a>bltabl <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../output/Ab_4-uniprot_blastx_sep.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-223"><a href="#cb45-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-224"><a href="#cb45-224" aria-hidden="true" tabindex="-1"></a>spgo <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"https://gannet.fish.washington.edu/seashell/snaps/uniprot_table_r2023_01.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb45-225"><a href="#cb45-225" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-226"><a href="#cb45-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-229"><a href="#cb45-229" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-230"><a href="#cb45-230" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(<span class="fu">head</span>(bltabl), <span class="at">options =</span> <span class="fu">list</span>(<span class="at">scrollX =</span> <span class="cn">TRUE</span>, <span class="at">scrollY =</span> <span class="st">"400px"</span>, <span class="at">scrollCollapse =</span> <span class="cn">TRUE</span>, <span class="at">paging =</span> <span class="cn">FALSE</span>))</span>
<span id="cb45-231"><a href="#cb45-231" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-232"><a href="#cb45-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-233"><a href="#cb45-233" aria-hidden="true" tabindex="-1"></a><span class="in">```{r spgo-table}</span></span>
<span id="cb45-234"><a href="#cb45-234" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(<span class="fu">head</span>(spgo), <span class="at">options =</span> <span class="fu">list</span>(<span class="at">scrollX =</span> <span class="cn">TRUE</span>, <span class="at">scrollY =</span> <span class="st">"400px"</span>, <span class="at">scrollCollapse =</span> <span class="cn">TRUE</span>, <span class="at">paging =</span> <span class="cn">FALSE</span>))</span>
<span id="cb45-235"><a href="#cb45-235" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-236"><a href="#cb45-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-237"><a href="#cb45-237" aria-hidden="true" tabindex="-1"></a><span class="in">```{r see}</span></span>
<span id="cb45-238"><a href="#cb45-238" aria-hidden="true" tabindex="-1"></a><span class="fu">datatable</span>(</span>
<span id="cb45-239"><a href="#cb45-239" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(bltabl, spgo,  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"V3"</span> <span class="ot">=</span> <span class="st">"Entry"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb45-240"><a href="#cb45-240" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">V1 =</span> <span class="fu">str_replace_all</span>(V1, </span>
<span id="cb45-241"><a href="#cb45-241" aria-hidden="true" tabindex="-1"></a>            <span class="at">pattern =</span> <span class="st">"solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed"</span>, <span class="at">replacement =</span> <span class="st">"Ab"</span>))</span>
<span id="cb45-242"><a href="#cb45-242" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb45-243"><a href="#cb45-243" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-244"><a href="#cb45-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-245"><a href="#cb45-245" aria-hidden="true" tabindex="-1"></a><span class="in">```{r join}</span></span>
<span id="cb45-246"><a href="#cb45-246" aria-hidden="true" tabindex="-1"></a>annot_tab <span class="ot">&lt;-</span></span>
<span id="cb45-247"><a href="#cb45-247" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(bltabl, spgo,  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"V3"</span> <span class="ot">=</span> <span class="st">"Entry"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb45-248"><a href="#cb45-248" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">V1 =</span> <span class="fu">str_replace_all</span>(V1, </span>
<span id="cb45-249"><a href="#cb45-249" aria-hidden="true" tabindex="-1"></a>            <span class="at">pattern =</span> <span class="st">"solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed"</span>, <span class="at">replacement =</span> <span class="st">"Ab"</span>))</span>
<span id="cb45-250"><a href="#cb45-250" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-251"><a href="#cb45-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-252"><a href="#cb45-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-255"><a href="#cb45-255" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-256"><a href="#cb45-256" aria-hidden="true" tabindex="-1"></a><span class="co"># Read dataset</span></span>
<span id="cb45-257"><a href="#cb45-257" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../output/blast_annot_go.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)  <span class="co"># Replace with the path to your dataset</span></span>
<span id="cb45-258"><a href="#cb45-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-259"><a href="#cb45-259" aria-hidden="true" tabindex="-1"></a><span class="co"># Select the column of interest</span></span>
<span id="cb45-260"><a href="#cb45-260" aria-hidden="true" tabindex="-1"></a>column_name <span class="ot">&lt;-</span> <span class="st">"Organism"</span>  <span class="co"># Replace with the name of the column of interest</span></span>
<span id="cb45-261"><a href="#cb45-261" aria-hidden="true" tabindex="-1"></a>column_data <span class="ot">&lt;-</span> dataset[[column_name]]</span>
<span id="cb45-262"><a href="#cb45-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-263"><a href="#cb45-263" aria-hidden="true" tabindex="-1"></a><span class="co"># Count the occurrences of the strings in the column</span></span>
<span id="cb45-264"><a href="#cb45-264" aria-hidden="true" tabindex="-1"></a>string_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(column_data)</span>
<span id="cb45-265"><a href="#cb45-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-266"><a href="#cb45-266" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to a data frame, sort by count, and select the top 10</span></span>
<span id="cb45-267"><a href="#cb45-267" aria-hidden="true" tabindex="-1"></a>string_counts_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(string_counts)</span>
<span id="cb45-268"><a href="#cb45-268" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(string_counts_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"String"</span>, <span class="st">"Count"</span>)</span>
<span id="cb45-269"><a href="#cb45-269" aria-hidden="true" tabindex="-1"></a>string_counts_df <span class="ot">&lt;-</span> string_counts_df[<span class="fu">order</span>(string_counts_df<span class="sc">$</span>Count, <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ]</span>
<span id="cb45-270"><a href="#cb45-270" aria-hidden="true" tabindex="-1"></a>top_10_strings <span class="ot">&lt;-</span> <span class="fu">head</span>(string_counts_df, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb45-271"><a href="#cb45-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-272"><a href="#cb45-272" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the top 10 most common strings using ggplot2</span></span>
<span id="cb45-273"><a href="#cb45-273" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(top_10_strings, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">reorder</span>(String, <span class="sc">-</span>Count), <span class="at">y =</span> Count, <span class="at">fill =</span> String)) <span class="sc">+</span></span>
<span id="cb45-274"><a href="#cb45-274" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">position =</span> <span class="st">"dodge"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb45-275"><a href="#cb45-275" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Top 10 Species hits"</span>,</span>
<span id="cb45-276"><a href="#cb45-276" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> column_name,</span>
<span id="cb45-277"><a href="#cb45-277" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Count"</span>) <span class="sc">+</span></span>
<span id="cb45-278"><a href="#cb45-278" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb45-279"><a href="#cb45-279" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb45-280"><a href="#cb45-280" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span>
<span id="cb45-281"><a href="#cb45-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-282"><a href="#cb45-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-283"><a href="#cb45-283" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-284"><a href="#cb45-284" aria-hidden="true" tabindex="-1"></a><span class="in">```{r go}</span></span>
<span id="cb45-285"><a href="#cb45-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-286"><a href="#cb45-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-287"><a href="#cb45-287" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"../output/blast_annot_go.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span>
<span id="cb45-288"><a href="#cb45-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-289"><a href="#cb45-289" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename the `Gene.Ontology..biological.process.` column to `Biological_Process`</span></span>
<span id="cb45-290"><a href="#cb45-290" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(data)[<span class="fu">colnames</span>(data) <span class="sc">==</span> <span class="st">"Gene.Ontology..biological.process."</span>] <span class="ot">&lt;-</span> <span class="st">"Biological_Process"</span></span>
<span id="cb45-291"><a href="#cb45-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-292"><a href="#cb45-292" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate the `Biological_Process` column into individual biological processes</span></span>
<span id="cb45-293"><a href="#cb45-293" aria-hidden="true" tabindex="-1"></a>data_separated <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(data<span class="sc">$</span>Biological_Process, <span class="at">split =</span> <span class="st">";"</span>))</span>
<span id="cb45-294"><a href="#cb45-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-295"><a href="#cb45-295" aria-hidden="true" tabindex="-1"></a><span class="co"># Trim whitespace from the biological processes</span></span>
<span id="cb45-296"><a href="#cb45-296" aria-hidden="true" tabindex="-1"></a>data_separated <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"^</span><span class="sc">\\</span><span class="st">s+|</span><span class="sc">\\</span><span class="st">s+$"</span>, <span class="st">""</span>, data_separated)</span>
<span id="cb45-297"><a href="#cb45-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-298"><a href="#cb45-298" aria-hidden="true" tabindex="-1"></a><span class="co"># Count the occurrences of each biological process</span></span>
<span id="cb45-299"><a href="#cb45-299" aria-hidden="true" tabindex="-1"></a>process_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(data_separated)</span>
<span id="cb45-300"><a href="#cb45-300" aria-hidden="true" tabindex="-1"></a>process_counts <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Biological_Process =</span> <span class="fu">names</span>(process_counts), <span class="at">Count =</span> <span class="fu">as.integer</span>(process_counts))</span>
<span id="cb45-301"><a href="#cb45-301" aria-hidden="true" tabindex="-1"></a>process_counts <span class="ot">&lt;-</span> process_counts[<span class="fu">order</span>(<span class="sc">-</span>process_counts<span class="sc">$</span>Count), ]</span>
<span id="cb45-302"><a href="#cb45-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-303"><a href="#cb45-303" aria-hidden="true" tabindex="-1"></a><span class="co"># Select the 20 most predominant biological processes</span></span>
<span id="cb45-304"><a href="#cb45-304" aria-hidden="true" tabindex="-1"></a>top_20_processes <span class="ot">&lt;-</span> process_counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, ]</span>
<span id="cb45-305"><a href="#cb45-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-306"><a href="#cb45-306" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a color palette for the bars</span></span>
<span id="cb45-307"><a href="#cb45-307" aria-hidden="true" tabindex="-1"></a>bar_colors <span class="ot">&lt;-</span> <span class="fu">rainbow</span>(<span class="fu">nrow</span>(top_20_processes))</span>
<span id="cb45-308"><a href="#cb45-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-309"><a href="#cb45-309" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a staggered vertical bar plot with different colors for each bar</span></span>
<span id="cb45-310"><a href="#cb45-310" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(top_20_processes<span class="sc">$</span>Count, <span class="at">names.arg =</span> <span class="fu">rep</span>(<span class="st">""</span>, <span class="fu">nrow</span>(top_20_processes)), <span class="at">col =</span> bar_colors,</span>
<span id="cb45-311"><a href="#cb45-311" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">max</span>(top_20_processes<span class="sc">$</span>Count) <span class="sc">*</span> <span class="fl">1.25</span>),</span>
<span id="cb45-312"><a href="#cb45-312" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Occurrences of the 20 Most Predominant Biological Processes"</span>, <span class="at">xlab =</span> <span class="st">"Biological Process"</span>, <span class="at">ylab =</span> <span class="st">"Count"</span>)</span>
<span id="cb45-313"><a href="#cb45-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-314"><a href="#cb45-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-315"><a href="#cb45-315" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a separate plot for the legend</span></span>
<span id="cb45-316"><a href="#cb45-316" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="st">"../output/GOlegend.png"</span>, <span class="at">width =</span> <span class="dv">800</span>, <span class="at">height =</span> <span class="dv">600</span>)</span>
<span id="cb45-317"><a href="#cb45-317" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))</span>
<span id="cb45-318"><a href="#cb45-318" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.new</span>()</span>
<span id="cb45-319"><a href="#cb45-319" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"center"</span>, <span class="at">legend =</span> top_20_processes<span class="sc">$</span>Biological_Process, <span class="at">fill =</span> bar_colors, <span class="at">cex =</span> <span class="dv">1</span>, <span class="at">title =</span> <span class="st">"Biological Processes"</span>)</span>
<span id="cb45-320"><a href="#cb45-320" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb45-321"><a href="#cb45-321" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-322"><a href="#cb45-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-323"><a href="#cb45-323" aria-hidden="true" tabindex="-1"></a><span class="in">```{r legend, fig.width = 100 ,fig.height = 100}</span></span>
<span id="cb45-324"><a href="#cb45-324" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"../output/GOlegend.png"</span>)</span>
<span id="cb45-325"><a href="#cb45-325" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-326"><a href="#cb45-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-327"><a href="#cb45-327" aria-hidden="true" tabindex="-1"></a><span class="fu"># Case Study - Oyster</span></span>
<span id="cb45-328"><a href="#cb45-328" aria-hidden="true" tabindex="-1"></a><span class="fu">## Navigating Annotation</span></span>
<span id="cb45-329"><a href="#cb45-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-330"><a href="#cb45-330" aria-hidden="true" tabindex="-1"></a>The following is a stepwise example or annotation of a gene set using UniProt::Swiss-Prot (reviewed) such that Gene Ontology terms can be associated with each gene.</span>
<span id="cb45-331"><a href="#cb45-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-332"><a href="#cb45-332" aria-hidden="true" tabindex="-1"></a>In this following chunk where the fasta file is downloaded the <span class="co">[</span><span class="ot">release</span><span class="co">](https://www.uniprot.org/help/release-statistics)</span> is noted and the file name is modified accordingly.</span>
<span id="cb45-333"><a href="#cb45-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-334"><a href="#cb45-334" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-335"><a href="#cb45-335" aria-hidden="true" tabindex="-1"></a>cd DRAFT_Funct_Enrich<span class="sc">/</span>annot</span>
<span id="cb45-336"><a href="#cb45-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-337"><a href="#cb45-337" aria-hidden="true" tabindex="-1"></a>curl <span class="sc">-</span>O https<span class="sc">:</span><span class="er">//</span>ftp.uniprot.org<span class="sc">/</span>pub<span class="sc">/</span>databases<span class="sc">/</span>uniprot<span class="sc">/</span>current_release<span class="sc">/</span>knowledgebase<span class="sc">/</span>complete<span class="sc">/</span>uniprot_sprot.fasta.gz</span>
<span id="cb45-338"><a href="#cb45-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-339"><a href="#cb45-339" aria-hidden="true" tabindex="-1"></a>mv uniprot_sprot.fasta.gz uniprot_sprot_r2023_02.fasta.gz</span>
<span id="cb45-340"><a href="#cb45-340" aria-hidden="true" tabindex="-1"></a>gunzip <span class="sc">-</span>k uniprot_sprot_r2023_02.fasta.gz</span>
<span id="cb45-341"><a href="#cb45-341" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-342"><a href="#cb45-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-343"><a href="#cb45-343" aria-hidden="true" tabindex="-1"></a>A protein blast database is then made.</span>
<span id="cb45-344"><a href="#cb45-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-345"><a href="#cb45-345" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-346"><a href="#cb45-346" aria-hidden="true" tabindex="-1"></a><span class="sc">/</span>home<span class="sc">/</span>shared<span class="sc">/</span>ncbi<span class="sc">-</span>blast<span class="dv">-2</span>.<span class="fl">11.0</span><span class="sc">+</span><span class="er">/</span>bin<span class="sc">/</span>makeblastdb \</span>
<span id="cb45-347"><a href="#cb45-347" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span><span class="cf">in</span> DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>uniprot_sprot_r2023_02.fasta \</span>
<span id="cb45-348"><a href="#cb45-348" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>dbtype prot \</span>
<span id="cb45-349"><a href="#cb45-349" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>out DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>uniprot_sprot_r2023_02</span>
<span id="cb45-350"><a href="#cb45-350" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-351"><a href="#cb45-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-352"><a href="#cb45-352" aria-hidden="true" tabindex="-1"></a>In a majority of cases you will want to annotate a gene set to get gene ontology information. If you are creating your own genome or transcriptome it should be rather straightforward to know what file to annotate. If using a widely studied system where there are publically available resources, it is advisable to use those as this is the best way to facilitate integration of data sets. In this case study we will be considering the Eastern oyster, (*Crassostrea virginica*) for which there is data at <span class="co">[</span><span class="ot">NCBI</span><span class="co">](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_002022765.2/)</span> and <span class="co">[</span><span class="ot">Ensembl Metazoa</span><span class="co">](http://metazoa.ensembl.org/Crassostrea_virginica_gca002022765v4/Info/Index)</span>. At NCBI there is both a GenBank and RefSeq assembly available.</span>
<span id="cb45-353"><a href="#cb45-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-354"><a href="#cb45-354" aria-hidden="true" tabindex="-1"></a>In order to know which of the numerous fasta files should annotated with gene ontology information one should think downstream (or look to files already generated) to the identifiers in genesets that would be subject to functional enrichment tests.</span>
<span id="cb45-355"><a href="#cb45-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-356"><a href="#cb45-356" aria-hidden="true" tabindex="-1"></a>The resulting fpkm count matrix for our case study is from an experiment where male and female oysters where exposed to low pH (and control) conditions. The count matrix is accessible <span class="co">[</span><span class="ot">here (csv)</span><span class="co">](https://github.com/epigeneticstoocean/2018_L18-adult-methylation/blob/main/analyses/gene_fpkm.csv)</span>. Hisat2/Stringtie was used to generate the count matrix with <span class="in">`GCF_002022765.2_C_virginica-3.0_genomic.gff`</span> formatting thus responsible for gene naming. Specifically the naming format is as follows <span class="in">`gene-LOC111099033,gene-LOC111099034,gene-LOC111099035`</span>.</span>
<span id="cb45-357"><a href="#cb45-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-358"><a href="#cb45-358" aria-hidden="true" tabindex="-1"></a>The following fasta was selected for annotation: <span class="in">`GCF_002022765.2_C_virginica-3.0_translated_cds.faa.gz`</span></span>
<span id="cb45-359"><a href="#cb45-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-360"><a href="#cb45-360" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-361"><a href="#cb45-361" aria-hidden="true" tabindex="-1"></a>cd DRAFT_Funct_Enrich<span class="sc">/</span>annot</span>
<span id="cb45-362"><a href="#cb45-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-363"><a href="#cb45-363" aria-hidden="true" tabindex="-1"></a>curl <span class="sc">-</span>O https<span class="sc">:</span><span class="er">//</span>ftp.ncbi.nlm.nih.gov<span class="sc">/</span>genomes<span class="sc">/</span>all<span class="sc">/</span>GCF<span class="sc">/</span><span class="dv">002</span><span class="sc">/</span><span class="dv">022</span><span class="sc">/</span><span class="dv">765</span><span class="sc">/</span>GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span><span class="sc">/</span>GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span>_translated_cds.faa.gz</span>
<span id="cb45-364"><a href="#cb45-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-365"><a href="#cb45-365" aria-hidden="true" tabindex="-1"></a>gunzip <span class="sc">-</span>k GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span>_translated_cds.faa.gz</span>
<span id="cb45-366"><a href="#cb45-366" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-367"><a href="#cb45-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-368"><a href="#cb45-368" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-369"><a href="#cb45-369" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">2</span> DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span>_translated_cds.faa</span>
<span id="cb45-370"><a href="#cb45-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-371"><a href="#cb45-371" aria-hidden="true" tabindex="-1"></a>echo <span class="st">"number of sequences"</span></span>
<span id="cb45-372"><a href="#cb45-372" aria-hidden="true" tabindex="-1"></a>grep <span class="sc">-</span>c  <span class="st">"&gt;"</span> DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span>_translated_cds.faa</span>
<span id="cb45-373"><a href="#cb45-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-374"><a href="#cb45-374" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-375"><a href="#cb45-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-376"><a href="#cb45-376" aria-hidden="true" tabindex="-1"></a><span class="in">```         </span></span>
<span id="cb45-377"><a href="#cb45-377" aria-hidden="true" tabindex="-1"></a><span class="in">&gt;lcl|NC_035780.1_prot_XP_022327646.1_1 [gene=LOC111126949] [db_xref=GeneID:111126949] [protein=UNC5C-like protein] [protein_id=XP_022327646.1] [location=join(30535..31557,31736..31887,31977..32565,32959..33204)] [gbkey=CDS]</span></span>
<span id="cb45-378"><a href="#cb45-378" aria-hidden="true" tabindex="-1"></a><span class="in">MTEVCYIWASSSTTVVICGIFFIVWRCFISIKKRASPLHGSSQQVCQTCQIEGHDFGEFQLSCRRQNTNVGYDLQGRRSD</span></span>
<span id="cb45-379"><a href="#cb45-379" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-380"><a href="#cb45-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-381"><a href="#cb45-381" aria-hidden="true" tabindex="-1"></a>This protein fasta is used as query for blast of uniprot_sprot database.</span>
<span id="cb45-382"><a href="#cb45-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-383"><a href="#cb45-383" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-384"><a href="#cb45-384" aria-hidden="true" tabindex="-1"></a><span class="sc">/</span>home<span class="sc">/</span>shared<span class="sc">/</span>ncbi<span class="sc">-</span>blast<span class="dv">-2</span>.<span class="fl">11.0</span><span class="sc">+</span><span class="er">/</span>bin<span class="sc">/</span>blastp \</span>
<span id="cb45-385"><a href="#cb45-385" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>query DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span>_translated_cds.faa \</span>
<span id="cb45-386"><a href="#cb45-386" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>db DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>uniprot_sprot_r2023_02 \</span>
<span id="cb45-387"><a href="#cb45-387" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>out DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>Cvir_transcds<span class="sc">-</span>uniprot_blastp.tab \</span>
<span id="cb45-388"><a href="#cb45-388" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>evalue <span class="fl">1E-20</span> \</span>
<span id="cb45-389"><a href="#cb45-389" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>num_threads <span class="dv">40</span> \</span>
<span id="cb45-390"><a href="#cb45-390" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>max_target_seqs <span class="dv">1</span> \</span>
<span id="cb45-391"><a href="#cb45-391" aria-hidden="true" tabindex="-1"></a><span class="sc">-</span>outfmt <span class="dv">6</span></span>
<span id="cb45-392"><a href="#cb45-392" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-393"><a href="#cb45-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-394"><a href="#cb45-394" aria-hidden="true" tabindex="-1"></a>Here is what the output file looks like, and at this point we want to get the UniProt Accession number for each gene</span>
<span id="cb45-395"><a href="#cb45-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-396"><a href="#cb45-396" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-397"><a href="#cb45-397" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">2</span> DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>Cvir_transcds<span class="sc">-</span>uniprot_blastp.tab</span>
<span id="cb45-398"><a href="#cb45-398" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-399"><a href="#cb45-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-400"><a href="#cb45-400" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eval=FALSE}</span></span>
<span id="cb45-401"><a href="#cb45-401" aria-hidden="true" tabindex="-1"></a>blast <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"DRAFT_Funct_Enrich/annot/Cvir_transcds-uniprot_blastp.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-402"><a href="#cb45-402" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-403"><a href="#cb45-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-404"><a href="#cb45-404" aria-hidden="true" tabindex="-1"></a>Convert fasta to tab</span>
<span id="cb45-405"><a href="#cb45-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-406"><a href="#cb45-406" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-407"><a href="#cb45-407" aria-hidden="true" tabindex="-1"></a>perl <span class="sc">-</span>e <span class="st">'$count=0; $len=0; while(&lt;&gt;) {s/</span><span class="sc">\r</span><span class="st">?</span><span class="sc">\n</span><span class="st">//; s/</span><span class="sc">\t</span><span class="st">/ /g; if (s/^&gt;//) { if ($. != 1) {print "</span><span class="sc">\n</span><span class="st">"} s/ |$/</span><span class="sc">\t</span><span class="st">/; $count++; $_ .= "</span><span class="sc">\t</span><span class="st">";} else {s/ //g; $len += length($_)} print $_;} print "</span><span class="sc">\n</span><span class="st">"; warn "</span><span class="sc">\n</span><span class="st">Converted $count FASTA records in $. lines to tabular format</span><span class="sc">\n</span><span class="st">Total sequence length: $len</span><span class="sc">\n\n</span><span class="st">";'</span> \</span>
<span id="cb45-408"><a href="#cb45-408" aria-hidden="true" tabindex="-1"></a>DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span>_translated_cds.faa <span class="sc">&gt;</span> DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span>_translated_cds.tab</span>
<span id="cb45-409"><a href="#cb45-409" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-410"><a href="#cb45-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-411"><a href="#cb45-411" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-412"><a href="#cb45-412" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">1</span> DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>GCF_002022765<span class="fl">.2</span>_C_virginica<span class="fl">-3.0</span>_translated_cds.tab</span>
<span id="cb45-413"><a href="#cb45-413" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-414"><a href="#cb45-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-417"><a href="#cb45-417" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb45-418"><a href="#cb45-418" aria-hidden="true" tabindex="-1"></a>cdsftab <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"DRAFT_Funct_Enrich/annot/GCF_002022765.2_C_virginica-3.0_translated_cds.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">FALSE</span>, <span class="at">row.names=</span><span class="cn">NULL</span>)</span>
<span id="cb45-419"><a href="#cb45-419" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-420"><a href="#cb45-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-421"><a href="#cb45-421" aria-hidden="true" tabindex="-1"></a>Now we can take the two data frames: A) blast output of taking protein fasta and comparing to uniprot_swiss-prot and B) a tabular version of same fasta file that has ID numbers of importance. Note this importance was determined based on what we want to use down stream.</span>
<span id="cb45-422"><a href="#cb45-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-423"><a href="#cb45-423" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb45-424"><a href="#cb45-424" aria-hidden="true" tabindex="-1"></a>g.spid <span class="ot">&lt;-</span> <span class="fu">left_join</span>(blast, cdsftab, <span class="at">by =</span> <span class="st">"V1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-425"><a href="#cb45-425" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">str_extract</span>(V2.y, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">[gene=)</span><span class="sc">\\</span><span class="st">w+"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb45-426"><a href="#cb45-426" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene, V11, V2.x) <span class="sc">%&gt;%</span></span>
<span id="cb45-427"><a href="#cb45-427" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">SPID =</span> <span class="fu">str_extract</span>(V2.x, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">|)[^</span><span class="sc">\\</span><span class="st">|]*(?=</span><span class="sc">\\</span><span class="st">|)"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb45-428"><a href="#cb45-428" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(gene, SPID, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>)</span>
<span id="cb45-429"><a href="#cb45-429" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-430"><a href="#cb45-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-431"><a href="#cb45-431" aria-hidden="true" tabindex="-1"></a>Let's break it down step by step:</span>
<span id="cb45-432"><a href="#cb45-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-433"><a href="#cb45-433" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>**`g.spid &lt;- left_join(blast, cdsftab, by = "V1")`** - This line is using the **`left_join()`** function from **`dplyr`** to merge the **`blast`** and **`cdsftab`** datasets by the column "V1". A left join retains all the rows in the **`blast`** data frame and appends the matching rows in the **`cdsftab`** data frame. If there is no match, the result is **`NA`**. The result of this operation is assigned to the **`g.spid`** object.</span>
<span id="cb45-434"><a href="#cb45-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-435"><a href="#cb45-435" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>**`mutate(gene = str_extract(V2.y, "(?&lt;=\\[gene=)\\w+"))`** - This line is using the **`mutate()`** function from **`dplyr`** to add a new column called "gene" to the data frame. The new column is created by extracting substrings from the "V2.y" column based on the given regular expression pattern **`"(?&lt;=\\[gene=)\\w+"`**. This regular expression matches and extracts any word (sequence of word characters, i.e., alphanumeric and underscore) that comes after "[gene=".</span>
<span id="cb45-436"><a href="#cb45-436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-437"><a href="#cb45-437" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>**`select(gene, V11, V2.x)`** - This line is using the **`select()`** function from **`dplyr`** to keep only the specified columns ("gene", "V11", and "V2.x") in the data frame.</span>
<span id="cb45-438"><a href="#cb45-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-439"><a href="#cb45-439" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>**`mutate(SPID = str_extract(V2.x, "(?&lt;=\\|)[^\\|]*(?=\\|)"))`** - Again, the **`mutate()`** function is used to add another new column named "SPID". This column is created by extracting substrings from the "V2.x" column. The regular expression **`"(?&lt;=\\|)[^\\|]*(?=\\|)"`** is designed to extract any character(s) that is/are surrounded by "\|" (pipe symbol). This is a common format for delimited strings.</span>
<span id="cb45-440"><a href="#cb45-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-441"><a href="#cb45-441" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>**`distinct(gene, SPID, .keep_all = TRUE)`** - This line is using the **`distinct()`** function from **`dplyr`** to remove duplicate rows based on the "gene" and "SPID" columns. The **`.keep_all = TRUE`** argument means that all other columns are also kept in the result, not just the "gene" and "SPID" columns.</span>
<span id="cb45-442"><a href="#cb45-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-443"><a href="#cb45-443" aria-hidden="true" tabindex="-1"></a>The resulting **`g.spid`** data frame should have unique rows with respect to the "gene" and "SPID" columns, and it should contain these two new columns, "gene" and "SPID", extracted from the original data based on specific string patterns.</span>
<span id="cb45-444"><a href="#cb45-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-445"><a href="#cb45-445" aria-hidden="true" tabindex="-1"></a>Now lets just write out SPIDs.</span>
<span id="cb45-446"><a href="#cb45-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-447"><a href="#cb45-447" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb45-448"><a href="#cb45-448" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(blast, cdsftab, <span class="at">by =</span> <span class="st">"V1"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-449"><a href="#cb45-449" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">str_extract</span>(V2.y, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">[gene=)</span><span class="sc">\\</span><span class="st">w+"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb45-450"><a href="#cb45-450" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene, V11, V2.x) <span class="sc">%&gt;%</span></span>
<span id="cb45-451"><a href="#cb45-451" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">SPID =</span> <span class="fu">str_extract</span>(V2.x, <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">|)[^</span><span class="sc">\\</span><span class="st">|]*(?=</span><span class="sc">\\</span><span class="st">|)"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb45-452"><a href="#cb45-452" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(gene, SPID, <span class="at">.keep_all =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-453"><a href="#cb45-453" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(SPID) <span class="sc">%&gt;%</span></span>
<span id="cb45-454"><a href="#cb45-454" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(<span class="at">file =</span> <span class="st">"DRAFT_Funct_Enrich/annot/SPID.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span></span>
<span id="cb45-455"><a href="#cb45-455" aria-hidden="true" tabindex="-1"></a> ) </span>
<span id="cb45-456"><a href="#cb45-456" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-457"><a href="#cb45-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-458"><a href="#cb45-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-459"><a href="#cb45-459" aria-hidden="true" tabindex="-1"></a>With a list of matching Swiss-Prot IDs, (technically UniProt Accession number) we can go back to https://www.uniprot.org and grab corresponding GO terms. This can be done via a web or using Python API.</span>
<span id="cb45-460"><a href="#cb45-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-461"><a href="#cb45-461" aria-hidden="true" tabindex="-1"></a>**Using Web**</span>
<span id="cb45-462"><a href="#cb45-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-463"><a href="#cb45-463" aria-hidden="true" tabindex="-1"></a>Using ID Mapping </span>
<span id="cb45-464"><a href="#cb45-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-465"><a href="#cb45-465" aria-hidden="true" tabindex="-1"></a><span class="al">![id](http://gannet.fish.washington.edu/seashell/snaps/Monosnap_RetrieveID_mapping__UniProt_2023-07-10_10-03-33.png)</span></span>
<span id="cb45-466"><a href="#cb45-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-467"><a href="#cb45-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-468"><a href="#cb45-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-469"><a href="#cb45-469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-470"><a href="#cb45-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-471"><a href="#cb45-471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-472"><a href="#cb45-472" aria-hidden="true" tabindex="-1"></a><span class="al">![finished](http://gannet.fish.washington.edu/seashell/snaps/Monosnap_Tool_results__UniProt_2023-07-10_10-05-07.png)</span></span>
<span id="cb45-473"><a href="#cb45-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-474"><a href="#cb45-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-475"><a href="#cb45-475" aria-hidden="true" tabindex="-1"></a>Now will customize columns to get GO IDs.</span>
<span id="cb45-476"><a href="#cb45-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-477"><a href="#cb45-477" aria-hidden="true" tabindex="-1"></a><span class="al">![custcol](http://gannet.fish.washington.edu/seashell/snaps/Monosnap_RetrieveID_mapping_results__UniProtKB__UniProt_2023-07-10_10-07-05.png)</span></span>
<span id="cb45-478"><a href="#cb45-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-479"><a href="#cb45-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-480"><a href="#cb45-480" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-481"><a href="#cb45-481" aria-hidden="true" tabindex="-1"></a>head <span class="sc">-</span><span class="dv">2</span> DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>uniprotGO.tab</span>
<span id="cb45-482"><a href="#cb45-482" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-483"><a href="#cb45-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-484"><a href="#cb45-484" aria-hidden="true" tabindex="-1"></a>Finally we can join table to get "LOCIDs" the notation for our DEGs, with GO terms.</span>
<span id="cb45-485"><a href="#cb45-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-486"><a href="#cb45-486" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb45-487"><a href="#cb45-487" aria-hidden="true" tabindex="-1"></a>go <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"DRAFT_Funct_Enrich/annot/uniprotGO.tab"</span>, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">row.names=</span><span class="cn">NULL</span>)</span>
<span id="cb45-488"><a href="#cb45-488" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-489"><a href="#cb45-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-490"><a href="#cb45-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-491"><a href="#cb45-491" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb45-492"><a href="#cb45-492" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(g.spid, go, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"SPID"</span> <span class="ot">=</span> <span class="st">"Entry"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb45-493"><a href="#cb45-493" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(gene,Gene.Ontology.IDs) <span class="sc">%&gt;%</span></span>
<span id="cb45-494"><a href="#cb45-494" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.table</span>(<span class="at">file =</span> <span class="st">"DRAFT_Funct_Enrich/annot/geneGO.txt"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">quote =</span> <span class="cn">FALSE</span></span>
<span id="cb45-495"><a href="#cb45-495" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb45-496"><a href="#cb45-496" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-497"><a href="#cb45-497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-498"><a href="#cb45-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-499"><a href="#cb45-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-500"><a href="#cb45-500" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash', eval=FALSE}</span></span>
<span id="cb45-501"><a href="#cb45-501" aria-hidden="true" tabindex="-1"></a>head DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>geneGO.txt</span>
<span id="cb45-502"><a href="#cb45-502" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb45-503"><a href="#cb45-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-504"><a href="#cb45-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-505"><a href="#cb45-505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-506"><a href="#cb45-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-507"><a href="#cb45-507" aria-hidden="true" tabindex="-1"></a>**Using API** </span>
<span id="cb45-508"><a href="#cb45-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-509"><a href="#cb45-509" aria-hidden="true" tabindex="-1"></a><span class="in">```{r engine='bash'}</span></span>
<span id="cb45-510"><a href="#cb45-510" aria-hidden="true" tabindex="-1"></a>python3 DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>uniprot<span class="sc">-</span>retrieval.py DRAFT_Funct_Enrich<span class="sc">/</span>annot<span class="sc">/</span>SPID.txt</span>
<span id="cb45-511"><a href="#cb45-511" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>